initial implementation of auth, login, signup for users

codedmart opened the pull request:
OK so this is a pretty good start, but I think you should either give some changes/suggestions on anything you think should be different or we should start setting new issues for other tasks off of this. At this point we can signup/login/logout. We can have protected routes now Here are some things that I think still need to happen that we should make new issues for.
- Need to send the userToken in localStorage with the api for auth ex: {"Authorization": "Token: hereIsAUserToken"}
- Related to above I did put the admin import log behind the protected routes as on example of using it so since we don't set a Authorization header it is broken with the pr
- Better erroring on the server. (Sounds like will be easier on servant 0.3)
- Any kind of error on login/signup forms (there are none now)
- Better errors on the client side so we can show errors on the login/signup form
  I am sure there are others but those are the main ones I can think of right now.

codedmart committed:
initial implementation of auth, login, signup for users

seanhess commented on server/Serials/Lib/Auth.hs:
```diff
@@ -0,0 +1,94 @@
+checkAuthToken :: Pool RethinkDBHandle -> TokenLookup
+checkAuthToken h token = do
+  users <- User.findByToken h $ decodeUtf8 $ token
+  case listToMaybe users of
```
`return $ isJust $ listToMaybe users`

codedmart replied:
Good call

seanhess commented on server/Serials/Lib/Auth.hs:
```diff
@@ -0,0 +1,94 @@
+  Nothing -> return Nothing
+  Just t -> do
+    users <- User.findByToken h t
+    case listToMaybe users of
```
`return $ listToMaybe users`

codedmart replied:
Good call

seanhess commented on server/Serials/Lib/Auth.hs:
```diff
@@ -0,0 +1,94 @@
+    Nothing -> return False
+    Just x -> return True
+
+-- Helpers should we move these to their respective models or to Helpers.hs or somewhere else?
```
In general? Depends on what it is. I'd leave it here until you needed it a second time.

codedmart replied:
Yeah just making a note there to remind myself. I can remove it if you want.

seanhess commented on server/Serials/Lib/Auth.hs:
```diff
@@ -0,0 +1,94 @@
+
+-- Helpers should we move these to their respective models or to Helpers.hs or somewhere else?
+textToByteString :: Text -> ByteString
+textToByteString = pack . unpack
```
`textToByteString = Data.Text.Encoding.encodeUtf8`

codedmart replied:
Good call

seanhess commented on server/Serials/Lib/Auth.hs:
```diff
@@ -0,0 +1,94 @@
+textToByteString = pack . unpack
+
+userPassword :: UserLogin -> Text
+userPassword = password
```
This function already exists. It's called `password` ? If you wanted to name it differently, why not name the record field name `userPassword`?

codedmart replied:
I will revisit but I had type errors when I did it inline where I needed it.

seanhess commented on server/Serials/Lib/Auth.hs:
```diff
@@ -0,0 +1,94 @@
+userPassword = password
+
+userHashedPassword :: User -> Text
+userHashedPassword = fromJust . User.hashedPassword
```
If you're using `fromJust` here, why not make it `Text` instead of `Maybe Text` in `User`?

codedmart replied:
Because we do not want to send the hashedPassword over the wire when we respond with the user.

seanhess replied:
Oh that's probably why you had a manual ToJSON instance too

seanhess commented on server/Serials/Lib/Crud.hs:
```diff
@@ -1,6 +1,6 @@
 {-# LANGUAGE OverloadedStrings #-}
 
-module Serials.Model.Crud where
+module Serials.Lib.Crud where
```
Lib is kind of a cop-out package name, don't you think?

codedmart replied:
No I didn't think so, but I don't care what it is called other then Crud does not feel like a model at all. So if you have other suggestions I am open. Unless you want it in Model, but that feels wrong to me.

seanhess replied:
`Serials.Model.Lib` ? It's full of db-model-related util stuff

seanhess commented on server/Serials/Model/User.hs:
```diff
@@ -0,0 +1,134 @@
+  , hashedPassword :: Maybe Text
+
+  , token :: Text
+  , admin :: Bool -- Should we have more permissions?
```
no this is good!

seanhess commented on server/Serials/Model/User.hs:
```diff
@@ -0,0 +1,134 @@
+} deriving (Show, Generic)
+
+instance FromJSON User where
+  parseJSON (Object v) = User
```
Why did you write manual JSON instances? did you want to control the field names or something?

codedmart replied:
Yes I don't want hashedPassword in the json when we respond with user/users

seanhess replied:
Cool. It may be easier to do this with composed types: `data SecureUser = SecureUser { user :: User, extra :: ... }`

codedmart replied:
Yeah if I understand what you are getting at right I am not sure I prefer that way. This way it straightforward to me, but we can discuss more.

seanhess commented on server/Serials/Model/User.hs:
```diff
@@ -0,0 +1,134 @@
+instance ToJSON User where
+  toJSON (User id firstName lastName email hashedPassword token admin created) = object $ catMaybes
+    [ ("id" .=) <$> pure id
+      , ("firstName" .=) <$> pure firstName
```
I think command-first style normally lines the commas up with the opening brace or bracket: 

```
[ ("id" .=) <$> pure id
, ("firstName" .=) <$> pure firstName
, ...
]
```

codedmart replied:
I couldn't remember :smile:

seanhess commented on server/Serials/Model/User.hs:
```diff
@@ -0,0 +1,134 @@
+          , admin = False
+          , created = created
+        }
+        case listToMaybe isEmail of
```
`listToMaybe` is the same as `Safe.headMay` right?

codedmart replied:
I believe so do you prefer headMay?

seanhess commented on server/Serials/Model/User.hs:
```diff
@@ -0,0 +1,134 @@
+        hashPass <- liftIO . hashPasswordUsingPolicy customHashPolicy . fromString $ unpack pass
+        isEmail <- findByEmail h $ U.email u
+        created <- liftIO getCurrentTime
+        token <- liftIO randomString
```
I think you're supposed to generate a token on login, rather than on insert.

codedmart replied:
I thought we could talk about this. I played it more as an api token that we keep around, but your suggestion is great if we don't need that.

seanhess commented on web/index.js:
```diff
@@ -12,12 +11,38 @@ var {Main} = require('./app/books/main')
+  }
+
+  setCurrentUser(user) {
+    updateLocalStorage('userToken', user.token)
```
Throw this into a client-side model? I'd recommend making a new model file. The current one is getting big.

codedmart replied:
OK I can change that and split the model file up.

codedmart replied:
Just to clarify did you mean to move the setCurrentUser out to the user model or something else?

seanhess replied:
The local storage access code. I'd wrap it in something.

On Friday, May 1, 2015, Brandon Martin notifications@github.com wrote:

> In web/index.js
> https://github.com/seanhess/serials/pull/18#discussion_r29543208:
> 
> >  class App extends React.Component {
> > +
> > -  constructor(props) {
> > -    super(props)
> > -    this.state = {currentUser: this.props.currentUser}
> > -  }
> >   +
> > -  componentDidMount() {
> > -    UserModel.checkAuth()
> > -    .then((user) => this.setState({currentUser: user}))
> > -  }
> >   +
> > -  setCurrentUser(user) {
> > -    updateLocalStorage('userToken', user.token)
> 
> Just to clarify did you mean to move the setCurrentUser out to the user
> model or something else?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/seanhess/serials/pull/18/files#r29543208.

codedmart replied:
Ok makes sense. Will do.

seanhess commented:
Wow, this is really great, I can't believe you got that done so fast! Thanks man!

Larger question: I think I was reading about there being new best practices for password hashing. (Like that salt + hash wasn't enough or something, I didn't get to the bottom). We should check that out later.

codedmart committed:
update some backend functions, separate models in client

codedmart commented:
I separated the model files on the client side and cleaned up some of the recommendations you had for the haskell stuff. I will do more on Monday and we can talk about some of the other points.

seanhess commented:
Awesome. Merge whenever you're ready.

codedmart committed:
add Crud.hs back in let user model handle localStorage

codedmart committed:
switch from token to jwt

codedmart commented:
OK a few last points. We can merge this, but I just want to point out a few things. I removed the token as something that we persist on the user. Switched the auth to use JWT. Right now I made it so the JWT token we send back expires after 24 hours just as an example but we can change that. I think this is a good start to work off of.
